File: frontend/src/types/route.ts
export interface FeatureScores {
  activity: number;
  businesses: number;
  emergency: number;
  time: number;
  isolation: number;
  surveillance: number;
  network: number;
  transit: number;
  lighting: number;
}

export interface AttentionZone {
  segment_start_index: number;
  segment_end_index: number;
  score: number;
  severity: "moderate" | "caution" | "attention";
  primary_reasons: string[];
  representative_coordinate: [number, number];
  coordinates: [number, number][];
}

export interface ExplanationDetail {
  status?: string;
  model_used?: string;
  headline?: string;
  summary?: string;
  why_recommended?: string[];
  tradeoffs?: string[];
  attention_summary?: string[];
}

export interface SegmentDetail {
  id: string;
  segment_index: number;
  length_m: number;
  segment_score: number;
  confidence: string;
  scores: FeatureScores;
  geometry: {
    type: string;
    coordinates: [number, number][];
  };
  reasons: string[];
  provenance?: Record<string, any>;
}

export interface RouteDetail {
  id: string;
  name: string;
  distance_m: number;
  duration_seconds: number;
  score: number;
  confidence: string;
  features: FeatureScores;
  attention_zones: AttentionZone[];
  segments: SegmentDetail[];
  geometry: {
    type: string;
    coordinates: [number, number][];
  };
  source?: string;
  validation?: {
    is_valid: boolean;
    status: string;
    reasons: string[];
    detour_ratio?: number;
    eta_ratio?: number;
  };
}

export interface RouteAnalysisResponse {
  request_id: string;
  travel_datetime: string;
  recommended_route_id: string;
  routes: RouteDetail[];
  explanation: string | ExplanationDetail;
  explanation_source?: string;
}

export interface RouteAnalysisRequest {
  origin_lat: number;
  origin_lng: number;
  destination_lat: number;
  destination_lng: number;
  origin_address?: string;
  destination_address?: string;
  travel_datetime: string;
  priority_safety: boolean;
}

File: backend/app/schemas/route.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any

class RouteRequestCreate(BaseModel):
    origin_lat: float = Field(..., description="Origin Latitude")
    origin_lng: float = Field(..., description="Origin Longitude")
    destination_lat: float = Field(..., description="Destination Latitude")
    destination_lng: float = Field(..., description="Destination Longitude")
    origin_address: Optional[str] = Field(None, description="Optional text address for origin")
    destination_address: Optional[str] = Field(None, description="Optional text address for destination")
    travel_datetime: str = Field(..., description="ISO 8601 travel date time")
    priority_safety: bool = Field(True, description="Prioritize contextual safety over travel time")

class FeatureScoresSchema(BaseModel):
    activity: float
    businesses: float
    emergency: float
    time: float
    isolation: float
    surveillance: float
    network: float
    transit: float
    lighting: float

class AttentionZoneSchema(BaseModel):
    segment_start_index: int
    segment_end_index: int
    segment_index: Optional[int] = None
    score: float
    severity: str # moderate, caution, attention
    primary_reasons: List[str]
    representative_coordinate: List[float]
    coordinates: List[List[float]]

class SegmentDetailSchema(BaseModel):
    id: str
    segment_index: int
    length_m: float
    segment_score: float
    confidence: str
    scores: FeatureScoresSchema
    geometry: Dict[str, Any]
    reasons: List[str]
    provenance: Optional[Dict[str, Any]] = None

class RouteDetailSchema(BaseModel):
    id: str
    name: str
    distance_m: float
    duration_seconds: float
    score: float
    confidence: str
    features: FeatureScoresSchema
    attention_zones: List[AttentionZoneSchema]
    segments: List[SegmentDetailSchema]
    geometry: Dict[str, Any]
    source: Optional[str] = "osrm"
    validation: Optional[Dict[str, Any]] = None

class RouteAnalysisResponse(BaseModel):
    request_id: str
    travel_datetime: str
    recommended_route_id: str
    routes: List[RouteDetailSchema]
    explanation: Any = "Route context analysis completed successfully."
    explanation_source: Optional[str] = "deterministic_fallback"
